crates/winscard-rs/src/scard.rs

```diff
@@ -0,0 +1,783 @@
+        }
+    }
+
+    fn general_authenticate(&mut self, cmd: Command<1024>) -> WinScardResult<Response> {
+        // NIST.SP.800-73-4, Part 2, Section 3.2.4
+        // PIV GENERAL AUTHENTICATE command
+        //      CLA  - 0x00 | 0x10 (command chaining)
+        //      INS  - 0x87
+        //      P1   - 0x07 - RSA
+        //      P2   - 0x9A - PIV Authentication Key
+        //      Data - Dynamic Authentication Template with Challenge inside
+        //
+        // There are many possible P1 and P2 values in this command, but our smart card only supports the RSA algorithm and data signing using the PIV Authentication Key
+
+        // NIST.SP.800-73-4, Part 1, Table 5
+        const RSA_ALGORITHM: u8 = 0x07;
+        // NIST.SP.800-73-4, Part 1, Table 4b
+        const PIV_AUTHENTICATION_KEY: u8 = 0x9A;
+
+        if cmd.p1 != RSA_ALGORITHM || cmd.p2 != PIV_AUTHENTICATION_KEY {
+            return Err(Error::new(
+                ErrorKind::UnsupportedFeature,
+                format!("provided algorithm or key reference isn't supported: got algorithm {}, expected 0x07; got key reference {}, expected 0x9A", cmd.p1, cmd.p2)
+            ));
+        }
+        let request = Tlv::from_bytes(cmd.data())?;
+        if request.tag() != &Tag::try_from(tlv_tags::DYNAMIC_AUTHENTICATION_TEMPLATE)?
+            || !request.value().is_constructed()
+        {
+            // wrong TLV request structure
+            return Err(Error::new(
+                ErrorKind::InvalidValue,
+                "TLV structure is invalid: wrong top-level tag structure".to_string(),
+            ));
+        }
+        let inner_tlv = match request.value() {
+            // we already know that the value is constructed at this point
+            Value::Primitive(_) => unreachable!(),
+            Value::Constructed(tlv_vec) => tlv_vec,
+        };
+        // to avoid constructing the tag on each iteration
+        let challenge_tag = Tag::try_from(tlv_tags::DAT_CHALLENGE)?;
+        let challenge = inner_tlv
+            .iter()
+            .find(|&tlv| tlv.tag() == &challenge_tag)
+            .ok_or(Error::new(
+                ErrorKind::InvalidValue,
+                "TLV structure is invalid: no challenge field is present in the request".to_string(),
+            ))?;
+        // Signature creation is described in NIST.SP.800-73-4, Part 2, Appendix A, Sections A.1-3 and Section A.4.1
+        let _challenge = match challenge.value() {
+            Value::Primitive(ref challenge) => challenge,
+            Value::Constructed(_) => {
+                // this tag must contain a primitive value
+                return Err(Error::new(
+                    ErrorKind::InvalidValue,
+                    "TLV structure is invalid: challenge field contains constructed value".to_string(),
+                ));
+            }
+        };
+        // TODO: actually sign the challenge
+        // let signed_challenge = sign_hashed_rsa(&self.auth_pk, challenge)?;
```

Comment:
This code was one of the problem places I found during debugging. I have a fix for it. I'll provide a separate pull request with all fixes